### PR TITLE
Do not depend on tree transformations for the add import refactoring 

### PR DESCRIPTION
**Pull request by sschaef**
This is a similar change to the rename refactoring. The overall idea is
to not depend on any tree transformations when an import is added to a
file. Theoretically, doing any transformations on the tree is the
optimal solution but only if the tree is a concrete syntax tree. The
scalac trees however are abstract syntax trees and therefore don't do
what we want. For the add import refactoring, they remove all the
formatting of the imports, especially removed new lines are a concern.

This commit therefore replaces the tree transformations by generating
textual code changes. This way, the refactoring is only doing a minimal
change, which in our case is to add one or multiple imports. A side
effect is that some functionality is lost but I consider that as a good
change. Formerly, the add import refactoring also did some minor
cleanups to the regions around imports. It added newlines, which means
that an import that got added to a code snippet like the following

```
  class A {
    val f = new File
  }
```

did result in this:

```
  import java.io.File

  class A {
    val f = new File
  }
```

Whereas after the refactorings of this commit the extra newline is no
longer printed:

```
  import java.io.File
  class A {
    val f = new File
  }
```

This may not look better but it makes the implementation easier because
we don't have to check if a newline already exists. Also, conceptually
it is the job of the organize imports refactoring to take care of such a
cleanup. The add import refactoring should only do a minimal change and
that is exactly what it is doing right now.

Fix #1002514

--------

@misto @mlangc Do you understand why I had to write this code:

```scala
var foundPackage: PackageDef = null
val findBestPackageForImports = {
  ...
  transformation[Tree, Tree] {
    case p @ PackageDef(_, stats) if isBestPlaceForImports(stats) =>
      if (foundPackage == null)
        foundPackage = p
      p
  }
}

// first try it at the top level to avoid traversing the complete AST
val insertLocation = findBestPackageForImports |> topdown(matchingChildren(findBestPackageForImports))

val addImportStatement = once(insertLocation) &> transformation { case _ ⇒
  splitImports(foundPackage, foundPackage.stats) match {
    case (p, imports, others) ⇒
```

I would like to write this code:
```scala
val findBestPackageForImports = {
  ...
  transformation[Tree, Tree] {
    case p @ PackageDef(_, stats) if isBestPlaceForImports(stats) =>
      splitImports(p, stats)
  }
}

// first try it at the top level to avoid traversing the complete AST
val insertLocation = findBestPackageForImports |> topdown(matchingChildren(findBestPackageForImports))

val addImportStatement = once(insertLocation) &> transformation {
  case (p, imports, others) ⇒
```

But the `once` block isn't doing what I would expect it to do. I expected, that it ensures that the transformation block in `findBestPackageForImports` is called only once, but it isn't and the found `PackageDef` is a different one in `addImportStatement`. Another problem is that I can't pass a transformation into `matchingChildren` that is not `Tree => Tree`

### CODE REVIEW

**Comment by misto** [comment]
Hm, does the `once` even have any effect there?

**Comment by misto** [comment]
You could try to wrap the transformation inside `findBestPackageForImports` in a `once` call, that might work. Right now, when you're using `once` in the outermost transformation, it shouldn't make a difference, because it's not applied more than once anyways.

**Comment by sschaef** [comment]
The `once` right now has no effect. Even if I put it into `findBestPackageForImports` it has no effect (or at least it doesn't do what I would expect it do to). Code works fine anyway but would be easier to understand without the `var`. Maybe it can be resolved by using an explicit Traverser to find the correct insertion point for the new import.

File: src/main/scala/scala/tools/refactoring/transformation/TreeTransformations.scala [hunk 1]
```diff
@@ -177,40 +177,80 @@ trait TreeTransformations extends Transformations with TreeFactory {
...
-        // The `pid` is invisible anyway, but by erasing its position we make sure
-        // it doesn't get any layout associated
-        val ignoredPid = p.pid setPos NoPosition
+    def insertBefore(pos: Position) = {
+      val lineNumber = pos.source.offsetToLine(pos.start)
+      val line = pos.source.lineToString(lineNumber)
+      val indent = line.takeWhile(Character.isWhitespace)
+      val insertPos = pos.source.lineToOffset(lineNumber)
+      TextChange(pos.source, insertPos, insertPos, importsAsSrc(indent) + "\n")
+    }
```
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/transformation/TreeTransformations.scala]: just to clarify. `pos: Position` is a position in original file. So when I have multiple changes applicable at the same position their are applied in order in list. (I cannot find it but I think that there was a piece which reverts the list of changes before application which has a sense)

**Comment by wpopielarski** [comment]
LGTM

**Comment by sschaef** [comment]
I have to merge this because I need to build upon. If anyone comes up with an idea about how to avoid the hack in the code, please speak up.

**Pull request merged**